host/logmux/sink.go

```diff
@@ -0,0 +1,608 @@
+	shutdownOnce sync.Once
+	shutdownCh   chan struct{}
+}
+
+func NewSyslogSink(sm *SinkManager, info *SinkInfo) (sink *SyslogSink, err error) {
+	cfg := &ct.SyslogSinkConfig{}
+	if err := json.Unmarshal(info.Config, cfg); err != nil {
+		return nil, err
+	}
+	var t *template.Template
+	var cache *lru.Cache
+	if cfg.Prefix != "" {
+		// Initialize the template cache
+		cache = lru.New(1000) // TODO(jpg): Consider configurable?
+		t, err = template.New("").Parse(cfg.Prefix)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &SyslogSink{
+		sm:         sm,
+		id:         info.ID,
+		url:        cfg.URL,
+		prefix:     cfg.Prefix,
+		cache:      cache,
+		template:   t,
+		cursor:     info.Cursor,
+		shutdownCh: make(chan struct{}),
+	}, nil
+}
+
+func (s *SyslogSink) Info() *SinkInfo {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	config, _ := json.Marshal(ct.SyslogSinkConfig{URL: s.url, Prefix: s.prefix})
+	return &SinkInfo{
+		ID:     s.id,
+		Kind:   ct.SinkKindSyslog,
+		Config: config,
+		Cursor: s.cursor,
+	}
+}
+
+func (s *SyslogSink) Connect() error {
+	u, err := url.Parse(s.url)
+	if err != nil {
+		return err
+	}
+	host, port, _ := net.SplitHostPort(u.Host)
+	if port == "" {
+		port = "514"
+	}
+	addr := net.JoinHostPort(host, port)
+	var conn net.Conn
+	switch u.Scheme {
+	case "syslog":
+		conn, err = net.Dial("tcp", addr)
+	case "syslog+tls":
+		tlsConfig := tlsconfig.SecureCiphers(&tls.Config{})
+		conn, err = tls.Dial("tcp", addr, tlsConfig)
+	default:
+		return fmt.Errorf("unknown protocol %s", u.Scheme)
+	}
+	if err != nil {
+		return err
+	}
+	s.conn = conn
+	return nil
+}
+
+func (s *SyslogSink) GetCursor(_ string) (*utils.HostCursor, error) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	return s.cursor, nil
+}
+
+// Parses JobID and ProcType from header ProcID field
+// Always returns JobID, returns ProcType if available
+func parseProcID(procID []byte) (string, string) {
+	procTypeID := strings.SplitN(string(procID), ".", 2)
+	if len(procTypeID) == 2 {
+		return procTypeID[1], procTypeID[0]
+	}
+	return procTypeID[0], ""
+}
+
+var msgSep = []byte{' '}
+
+func (s *SyslogSink) Write(m message) error {
+	if s.template != nil {
+		// Lookup rendered template from cache
+		var prefix []byte
+		if cached, ok := s.cache.Get(m.Message.ProcID); ok {
+			if p, ok := cached.([]byte); ok {
+				prefix = p
+			}
+		}
+		// If not in the cache execute the template and cache the result
+		if prefix == nil {
+			jobID, _ := parseProcID(m.Message.ProcID)
+			job := s.sm.state.GetJob(jobID)
+			if job != nil && job.Job != nil {
+				var buf bytes.Buffer
+				if err := s.template.Execute(&buf, job.Job); err != nil {
+					return err
+				}
+				prefix = buf.Bytes()
+				s.cache.Add(jobID, prefix)
+			}
+		}
+		// If the generated/cached prefix isn't 0 length then modify the message body
+		if len(prefix) != 0 {
+			m.Message.Msg = bytes.Join([][]byte{prefix, m.Message.Msg}, msgSep)
+		}
+	}
+	_, err := s.conn.Write(rfc6587.Bytes(m.Message))
```

Comment:
I think making it available to the prefix formatter is probably the best idea. That way it can be used to populate tags etc for system that support structured parsing. Right now it is accessed via the job meta but replacing the job with something that just contains meta, app name, env is probably what I am going to do (actually have already done just haven't rebased/pushed yet).